gcc/rust/backend/rust-compile-intrinsic.cc

```diff
@@ -93,20 +93,46 @@ prefetch_write_data (Context *ctx, TyTy::FnType *fntype)
   return prefetch_data_handler (ctx, fntype, Prefetch::Write);
 }
 
+static tree
+atomic_store_handler_inner (Context *ctx, TyTy::FnType *fntype, int ordering);
+
+static inline std::function<tree (Context *, TyTy::FnType *)>
+atomic_store_handler (int ordering)
+{
+  return [ordering] (Context *ctx, TyTy::FnType *fntype) {
+    return atomic_store_handler_inner (ctx, fntype, ordering);
+  };
+}
+
+static inline tree
+sorry_handler (Context *ctx, TyTy::FnType *fntype)
+{
+  rust_sorry_at (fntype->get_locus (), "intrinsic %qs is not yet implemented",
+		 fntype->get_identifier ().c_str ());
+
+  return error_mark_node;
+}
+
 static const std::map<std::string,
 		      std::function<tree (Context *, TyTy::FnType *)>>
   generic_intrinsics = {
-    {"offset", &offset_handler},
-    {"size_of", &sizeof_handler},
-    {"transmute", &transmute_handler},
-    {"rotate_left", &rotate_left_handler},
-    {"rotate_right", &rotate_right_handler},
-    {"wrapping_add", &wrapping_add_handler},
-    {"wrapping_sub", &wrapping_sub_handler},
-    {"wrapping_mul", &wrapping_mul_handler},
-    {"copy_nonoverlapping", &copy_nonoverlapping_handler},
-    {"prefetch_read_data", &prefetch_read_data},
-    {"prefetch_write_data", &prefetch_write_data},
+    {"offset", offset_handler},
+    {"size_of", sizeof_handler},
+    {"transmute", transmute_handler},
+    {"rotate_left", rotate_left_handler},
+    {"rotate_right", rotate_right_handler},
+    {"wrapping_add", wrapping_add_handler},
+    {"wrapping_sub", wrapping_sub_handler},
+    {"wrapping_mul", wrapping_mul_handler},
+    {"copy_nonoverlapping", copy_nonoverlapping_handler},
+    {"prefetch_read_data", prefetch_read_data},
+    {"prefetch_write_data", prefetch_write_data},
+    {"atomic_load", sorry_handler},
+    {"atomic_store_seqcst", atomic_store_handler (__ATOMIC_SEQ_CST)},
+    {"atomic_store_release", atomic_store_handler (__ATOMIC_RELEASE)},
+    {"atomic_store_relaxed", atomic_store_handler (__ATOMIC_RELAXED)},
+    {"atomic_store_unordered", sorry_handler},
+    // FIXME: Are _unordered and _relaxed the same?
```

Comment:
Went hunting a bit since I haven't heard of this.

https://llvm.org/docs/Atomics.html#unordered
Looks like no, `unordered` is even weaker than `relaxed` and comes from Java rather than C/C++.

https://llvm.org/docs/LangRef.html#atomic-memory-ordering-constraints
`This ordering cannot be specified for read-modify-write operations; it is not strong enough to make them atomic in any interesting way.`

That said, since the guarantees of `unordered` are strictly weaker than `relaxed` it should be fine to map this to `__ATOMIC_RELAXED`.

Reply:
Heh bjorn beat me to it :P